spmd/compiler/fusion.py

```diff
@@ -0,0 +1,182 @@
+from copy import deepcopy
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Iterable, List, Optional
+
+import torch
+import torch.distributed as dist
+import torch.fx as fx
+from .graph_utils import (
+    OP,
+    get_node_tensor_numel,
+    get_output_node,
+    graph_cleanup,
+    pretty_print_graph,
+)
+
+
+# enum for the supported fusion comm types
+class Comm_Type(str, Enum):
+    allreduce = "allreduce_"
+    allgather = "allgather_"
+    broadcast = "broadcast_"
+    reducescatter = "reduce_scatter_"
+    scatter = "scatter_"
+
+
+@dataclass
+class FusionElement:
+    comm_type: Optional[Comm_Type] = None
+    node_list: List = field(default_factory=lambda: [])  # type: ignore
```

Comment:
Unfortunately not. It will complain that object does not have property .prev etc. (node properties).
  I also tried Optional[List[fx.Node]] but  ultimately had to go back to # type: ignore b/c it doesn't otherwise understand that its a list of nodes, that may or may not be populated.